### PR TITLE
Enable SSL certificate check in http helper

### PR DESCRIPTION
**Pull request by klyonrad**
# Summary

This removes bypassing the SSL certificate check in the custom http helper.
# Test Plan
- `vagrant provision`
- `cucumber`

### CODE REVIEW

File: lib/dkdeploy/php/helpers/http.rb [hunk 1]
```diff
@@ -41,7 +41,7 @@ def http_get_with_redirect(url, limit = 5)
           http.read_timeout = fetch :http_read_timeout
           if url.scheme == 'https'
             http.use_ssl = true
-            http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+            http.verify_mode = OpenSSL::SSL::VERIFY_PEER
```
Comment (klyonrad) [inline lib/dkdeploy/php/helpers/http.rb]: `VERIFY_PEER` is the default and this line could actually be removed. But this way the change is more explicit

File: config/vm/cookbooks/dkdeploy-php/recipes/default.rb [hunk 2]
```diff
@@ -75,3 +76,16 @@
...
+my_key_path = '/var/www/dkdeploy.key'
+my_cert_path = '/var/www/dkdeploy.pem'
+
+ssl_certificate 'dkdeploy-php.dev' do
+  key_path my_key_path
+  key_mode 00755
+  cert_path my_cert_path
+  domain 'dkdeploy-php.dev'
+  organization 'dkdeploy'
+  email 'luka.luedicke@dkd.de'
```
Comment (ncreuschling) [inline config/vm/cookbooks/dkdeploy-php/recipes/default.rb]: Please remove personal reference. Replace with example.com address.

File: config/vm/cookbooks/dkdeploy-php/recipes/default.rb [hunk 3]
```diff
@@ -75,3 +76,16 @@
   mode '0770'
   action :create
 end
+
+my_key_path = '/var/www/dkdeploy.key'
```
Comment (ncreuschling) [inline config/vm/cookbooks/dkdeploy-php/recipes/default.rb]: Make customizable or move inline.

File: config/vm/cookbooks/dkdeploy-php/recipes/default.rb (hunk 2 above)
Comment (ncreuschling) [reply config/vm/cookbooks/dkdeploy-php/recipes/default.rb]: Dito

File: config/vm/cookbooks/dkdeploy-php/templates/default/web_app.conf.erb [hunk 4]
```diff
@@ -26,3 +26,38 @@
...
+  </Directory>
+
+  <Directory />
+  Options FollowSymLinks
+  AllowOverride None
+  </Directory>
+
+  <IfModule mod_ssl.c>
+    SSLEngine on
+    SSLCertificateFile /var/www/dkdeploy.pem
```
Comment (ncreuschling) [inline config/vm/cookbooks/dkdeploy-php/templates/default/web_app.conf.erb]: Use variable from recipe.

**Review by ncreuschling — CHANGES_REQUESTED** [review]

File: config/vm/cookbooks/dkdeploy-php/metadata.rb [hunk 5]
```diff
@@ -10,3 +10,4 @@
 depends 'apache2', '~> 3.2'
 depends 'php', '~> 1.9'
 depends 'apt', '~> 3.0'
+depends 'ssl_certificate'
```
Comment (websi) [inline config/vm/cookbooks/dkdeploy-php/metadata.rb]: Please add version number for `ssl_certificate`

File: features/php.feature [hunk 6]
```diff
@@ -42,3 +42,19 @@ Feature: Test tasks for namespace 'php'
 		And I extend the development capistrano configuration from the fixture file server_with_faulty_domain_configuration.rb
 		When I run `cap dev php:clear_opcache`
 		Then the exit status should not be 0
+
+	Scenario: Clear APC via HTTPS
+		Given a directory named "temp"
+		Given I extend the development capistrano configuration from the fixture file server_with_ssl.rb
+		Given I set relative to current path the environment variable "SSL_CERT_FILE" to "/tmp/aruba/temp/dkdeploy.dkdeploy-php.dev.pem"
```
Comment (websi) [inline features/php.feature]: Why `relative to current`?
Can you remove path `/tmp/aruba/temp/`?

File: features/php.feature [hunk 7]
```diff
@@ -42,3 +42,19 @@ Feature: Test tasks for namespace 'php'
...
+		When I run `cap dev utils:download_file[../../../dkdeploy.pem]`
```
Comment (websi) [inline features/php.feature]: Is this necessary?

File: lib/dkdeploy/php/helpers/http.rb (hunk 1 above)
Comment (websi) [reply lib/dkdeploy/php/helpers/http.rb]: Maybe implement an configuration option for this.

**Review by websi — CHANGES_REQUESTED** [review]

File: features/php.feature (hunk 6 above)
Comment (klyonrad) [reply features/php.feature]: We need the absolute path to the certificate file and string interpolation is possible in the cucumber step.  The RegExp match data alwasy escapes the special charachter for string interpolation

Comment (klyonrad) [reply features/php.feature]: > Can you remove path /tmp/aruba/temp/?

Yeah, I am using the `test_app_path` method now in the step definition

File: features/php.feature (hunk 7 above)
Comment (klyonrad) [reply features/php.feature]: The certificate needs to get copied from the test VM in some way

File: lib/dkdeploy/php/helpers/http.rb [hunk 8]
```diff
@@ -41,7 +41,7 @@ def http_get_with_redirect(url, limit = 5)
           http.read_timeout = fetch :http_read_timeout
           if url.scheme == 'https'
             http.use_ssl = true
-            http.verify_mode = OpenSSL::SSL::VERIFY_NONE
+            http.verify_mode = OpenSSL::SSL::VERIFY_NONE if disable_ssl_check
```
Comment (klyonrad) [inline lib/dkdeploy/php/helpers/http.rb]: @websi I guess you would like that configurable in the server properties and how should it be called?

**Comment by klyonrad** [comment]
made some improvements, updated changelog

File: lib/dkdeploy/php/helpers/http.rb [hunk 9]
```diff
@@ -18,20 +18,22 @@ def call_file_on_server(filename, server)
...
           # Use server configuration, if exists
           web_server_port = server.fetch(:web_server_port) if server.properties.respond_to?(:web_server_port)
+          ssl_verify_mode = server.properties.respond_to?(:ssl_verify_mode) ? server.fetch(:ssl_verify_mode) : OpenSSL::SSL::VERIFY_PEER
 
           url = URI.parse("#{domain_scheme}://#{domain}").merge("/#{filename}")
           url.port = web_server_port
           info "Call URL #{url}"
 
-          http_get_with_redirect url
+          http_get_with_redirect url: url, verify_mode: ssl_verify_mode
```
Comment (websi) [inline lib/dkdeploy/php/helpers/http.rb]: Is this a breaking change?
If yes, we need to document this breaking change.
Or can we use `http_get_with_redirect(url, limit = 5, verify_mode: OpenSSL::SSL::VERIFY_PEER)`?

File: lib/dkdeploy/php/helpers/http.rb [hunk 10]
```diff
@@ -18,20 +18,22 @@ def call_file_on_server(filename, server)
...
         end
 
         # Sends a get request that handles redirects
         #
         # @param url [URI]
         # @param limit [Integer] defines how many redirects are allowed
+        # @param verify_mode [OpenSSL::SSL::verify_mode] ssl verify mode, setting to VERIFY_NONE disables certificate check
         # @return [NET::HTTPResponse]
-        def http_get_with_redirect(url, limit = 5)
+        def http_get_with_redirect(url:, limit: 5, verify_mode: OpenSSL::SSL::VERIFY_PEER)
```
Comment (websi) [inline lib/dkdeploy/php/helpers/http.rb:36]: Is this a breaking change?
If yes, we need to document this breaking change.
Or can we use http_get_with_redirect(url, limit = 5, verify_mode: OpenSSL::SSL::VERIFY_PEER)?

Comment (klyonrad) [reply lib/dkdeploy/php/helpers/http.rb:36]: Yes, it's a breaking change, I documented that in the changelog in the section _Changed_

File: config/vm/cookbooks/dkdeploy-php/recipes/default.rb [hunk 11]
```diff
@@ -75,3 +81,13 @@
...
   action :create
 end
+
+ssl_certificate 'dkdeploy-php.dev' do
+  key_path ssl_key_path
+  key_mode 00755
+  cert_path ssl_cert_path
+  domain 'dkdeploy-php.dev'
+  organization 'dkdeploy'
+  email 'offical@example.com'
```
Comment (websi) [inline config/vm/cookbooks/dkdeploy-php/recipes/default.rb:91]: Please use no example email address.

**Review by websi — DISMISSED** [review]

**Review by websi — APPROVED** [review]
@ncreuschling can you take a look of this

File: .rubocop.yml [hunk 12]
```diff
@@ -1,4 +1,5 @@
 AllCops:
+  TargetRubyVersion: 2.1
```
Comment (ncreuschling) [inline .rubocop.yml]: We will target v2.2 now. And we should add it to `dkdeploy-core` and `dkdeploy-typo3-cms` aswell. I can take care of it.

**Review by ncreuschling — APPROVED** [review]

**Comment by ncreuschling** [comment]
Please squash and rebase. Then we'll merge it.

**Pull request merged**